Add MutationRecord Properties

cowtoolz opened the pull request:
### Description

Add documentation for all properties of the `MutationRecord` interface (which do not exist at present).

Documented with examples
- addedNodes
- attributeName
- nextSibling
- oldValue
- previousSibling
- removedNodes
- target
- type

Documented, but without examples
- attributeNamespace

### Motivation
Working on a project that relies heavily on `MutationObserver` and manipulation of nodes leveraging `MutationRecord`; the documentation regarding `MutationRecord`  was not useful. Decided to pause working on my project to flesh out the public documentation on the interface and its methods.

cowtoolz committed:
Add all new content- methods & rewrite of main index

Adds the following methods
    addedNodes()
    attributeName()
    attributeNamespace()
    nextSibling()
    oldValue()
    previousSibling()
    removedNodes()
    target()
    type()

Josh-Cena committed:
Merge branch 'main' into add-MutationRecord-methods

mdn-bot committed:
chore: auto-fix Markdownlint issues

wbamberg commented:
Thanks for this PR, @weebney ! I'll take a look in the next few days...

cowtoolz committed:
Merge pull request #2 from weebney/markdownlint-auto-cleanup

Markdownlint auto-cleanup

cowtoolz committed:
Merge branch 'main' into add-MutationRecord-methods

wbamberg reviewed (changes requested):
Thanks for your PR, @weebney ! This is quite an important API to have missing pages, so I really appreciate your work.

I've only reviewed the first half at this point, but ran out of time for today. I'll add some more comments tomorrow - it's probably worth you waiting until I've finished before you make  updates. But please ask if anything's unclear :).

cowtoolz committed:
Update files/en-us/web/api/mutationrecord/addednodes/index.md

Co-authored-by: wbamberg <will@bootbonnet.ca>

cowtoolz committed:
Commit suggestions

Update files/en-us/web/api/mutationrecord/addednodes/index.md

Update files/en-us/web/api/mutationrecord/attributename/index.md

Update files/en-us/web/api/mutationrecord/index.md

Update files/en-us/web/api/mutationrecord/oldvalue/index.md

Update files/en-us/web/api/mutationrecord/attributename/index.md

Update files/en-us/web/api/mutationrecord/attributename/index.md

Co-Authored-By: wbamberg <will@bootbonnet.ca>

cowtoolz committed:
Update files/en-us/web/api/mutationrecord/addednodes/index.md

Co-authored-by: wbamberg <will@bootbonnet.ca>

cowtoolz committed:
Update files/en-us/web/api/mutationrecord/attributename/index.md

Co-authored-by: wbamberg <will@bootbonnet.ca>

cowtoolz committed:
Update files/en-us/web/api/mutationrecord/index.md

Co-authored-by: wbamberg <will@bootbonnet.ca>

cowtoolz committed:
Update files/en-us/web/api/mutationrecord/oldvalue/index.md

Co-authored-by: wbamberg <will@bootbonnet.ca>

cowtoolz committed:
Update files/en-us/web/api/mutationrecord/attributename/index.md

Co-authored-by: wbamberg <will@bootbonnet.ca>

cowtoolz committed:
Update files/en-us/web/api/mutationrecord/attributename/index.md

Co-authored-by: wbamberg <will@bootbonnet.ca>

cowtoolz committed:
Merge branch 'add-MutationRecord-methods' of https://github.com/weebney/content into add-MutationRecord-methods

cowtoolz committed:
Semantic & formatting changes; add live samples

Update to live samples:
addedNodes
removedNodes
attributeName

weebney commented:
You're helping me learn a lot here and I deeply appreciate both your patience & thoroughness. I've updated all pages to reflect your criticism and I am in the process of writing live samples demonstrating each *property* ;) 

Please do check out the currently committed live samples in addedNodes, removedNodes, and attributeName to review them for code quality noting my general freshness to programming.

dependabot[bot] committed:
chore(deps): bump tj-actions/changed-files from 34.1.1 to 34.3.2

Bumps [tj-actions/changed-files](https://github.com/tj-actions/changed-files) from 34.1.1 to 34.3.2.
- [Release notes](https://github.com/tj-actions/changed-files/releases)
- [Changelog](https://github.com/tj-actions/changed-files/blob/main/HISTORY.md)
- [Commits](https://github.com/tj-actions/changed-files/compare/v34.1.1...v34.3.2)

---
updated-dependencies:
- dependency-name: tj-actions/changed-files
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
chore(deps): bump husky from 8.0.1 to 8.0.2

Bumps [husky](https://github.com/typicode/husky) from 8.0.1 to 8.0.2.
- [Release notes](https://github.com/typicode/husky/releases)
- [Commits](https://github.com/typicode/husky/compare/v8.0.1...v8.0.2)

---
updated-dependencies:
- dependency-name: husky
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
chore(deps): bump @mdn/yari from 1.22.4 to 1.23.0

Bumps [@mdn/yari](https://github.com/mdn/yari) from 1.22.4 to 1.23.0.
- [Release notes](https://github.com/mdn/yari/releases)
- [Changelog](https://github.com/mdn/yari/blob/main/docs/npm-releases.md)
- [Commits](https://github.com/mdn/yari/compare/v1.22.4...v1.23.0)

---
updated-dependencies:
- dependency-name: "@mdn/yari"
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

cowtoolz committed:
Merge pull request #5 from weebney/dependabot/npm_and_yarn/mdn/yari-1.23.0

chore(deps): bump @mdn/yari from 1.22.4 to 1.23.0

cowtoolz committed:
Merge pull request #4 from weebney/dependabot/npm_and_yarn/husky-8.0.2

chore(deps): bump husky from 8.0.1 to 8.0.2

cowtoolz committed:
Merge pull request #3 from weebney/dependabot/github_actions/tj-actions/changed-files-34.3.2

chore(deps): bump tj-actions/changed-files from 34.1.1 to 34.3.2

cowtoolz committed:
Merge branch 'main' into add-MutationRecord-methods

cowtoolz committed:
Move table to def list; add live sample for oldValue

cowtoolz committed:
Fix errors in sample

cowtoolz committed:
Rewording

cowtoolz committed:
New examples + cleanup

wbamberg commented on files/en-us/web/api/mutationrecord/target/index.md:
````diff
@@ -0,0 +1,45 @@
+In this code snippet, the `observer` will log to the console the tag name of every element with an observed mutation with the {{domxref("MutationObserver.observe()")}} property.
+
+```js
+const observer = new MutationObserver(function(mutations) {
+    mutations.forEach(function(mutation) {
+        console.log(mutation.target.tagName);
+    });
+});
````
```suggestion
const observer = new MutationObserver((mutations) => {
    mutations.forEach((mutation) => {
        console.log(mutation.target.tagName);
    });
});
```

wbamberg commented on files/en-us/web/api/mutationrecord/target/index.md:
```diff
@@ -0,0 +1,45 @@
+
+## Examples
+
+In this code snippet, the `observer` will log to the console the tag name of every element with an observed mutation with the {{domxref("MutationObserver.observe()")}} property.
```
```suggestion
In this code snippet, the `observer` will log to the console the tag name of every element with an observed mutation.
```

(at least, I didn't understand "with the {{domxref("MutationObserver.observe()")}} property")

wbamberg commented on files/en-us/web/api/mutationrecord/type/index.md:
```diff
@@ -0,0 +1,66 @@
+
+{{APIRef("DOM")}}
+
+The {{domxref("MutationRecord")}} property **`type`** is the type of the {{domxref("MutationRecord")}} observed by a {{domxref("MutationObserver")}}.
```
```suggestion
The {{domxref("MutationRecord")}} read-only property **`type`** is the type of the {{domxref("MutationRecord")}} observed by a {{domxref("MutationObserver")}}.
```

wbamberg commented on files/en-us/web/api/mutationrecord/type/index.md:
```diff
@@ -0,0 +1,66 @@
+
+## Value
+
+The property is set to the type of the mutated node as a string. The value can be one of the following:
```
```suggestion
The property is set to the type of the mutation as a string. The value can be one of the following:
```

wbamberg commented on files/en-us/web/api/mutationrecord/type/index.md:
```diff
@@ -0,0 +1,66 @@
+
+The property is set to the type of the mutated node as a string. The value can be one of the following:
+
+- The value will be `attributes` if the mutation to the target node of the was an attribute mutation, {{domxref("Attr")}}. This mutated node type can appear if the "attribute" parameter of {{domxref("MutationObserver.observe()")}} is set to true or the `attributeFilter` parameter is in use.
+
+- The value will be `childList` if the mutation to the target node was a mutation to the tree of nodes, i.e. the `childList` paremeter of {{domxref("MutationObserver.observe()")}} is set to true.
+
+- The value will be `characterData` if it was called on a mutation to a {{domxref("CharacterData")}} node.
```
```suggestion
- `attributes` if the mutation was an attribute mutation.

- `characterData` if it was a mutation to a {{domxref("CharacterData")}} node.

- `childList` if the mutation a mutation to the tree of nodes.
```

wbamberg commented on files/en-us/web/api/mutationrecord/type/index.md:
```diff
@@ -0,0 +1,66 @@
+
+## Examples
+
+This code snippet is used to observe changes in a userlist: the userlist is DOM element that functions as a list of all users currently connected to a server (in this example, consider it is a `<ul>`). The userlist itself contains child elements (i.e. `<li>`), each containing an attribute associated with the username the user status.
```
Consider a live sample based on this pen: https://codepen.io/willbamberg/pen/wvXPoPB ?

wbamberg reviewed (changes requested):
Sorry to be so slow to review this PR, @weebney ! I finally went through it all. Please ask if you have any q's.

cowtoolz committed:
Apply suggestions from code review

Co-authored-by: wbamberg <will@bootbonnet.ca>

cowtoolz committed:
Merge branch 'add-MutationRecord-methods' of https://github.com/weebney/content into add-MutationRecord-methods

cowtoolz committed:
Replace & update live samples; general cleanup

cowtoolz committed:
Merge branch 'mdn:main' into add-MutationRecord-methods

cowtoolz committed:
Fix markdown

cowtoolz committed:
Merge branch 'add-MutationRecord-methods' of https://github.com/weebney/content into add-MutationRecord-methods

wbamberg commented on files/en-us/web/api/mutationrecord/type/index.md:
```diff
@@ -0,0 +1,109 @@
+<button id="reset">Reset</button>
+
+<pre id= "log">Mutation type:</pre>
+<div id= "target" class="my-class"></div>
```
```suggestion
<div id= "target"></div>
```

(sorry, my fault!)

wbamberg committed:
Update files/en-us/web/api/mutationrecord/type/index.md

wbamberg commented on files/en-us/web/api/mutationrecord/previoussibling/index.md:
```diff
@@ -0,0 +1,90 @@
+    if (record.type === "childList") {
+      for (const newNode of record.addedNodes) {
+        counter.textContent = `Node count: ${target.children.length}`;
+        log.textContent = `Previous sibling of added node: ${newNode.previousSibling.textContent}`;
```
```suggestion
        log.textContent = `Previous sibling of added node: ${newNode.previousSibling?.textContent}`;
```

(because the first time `previousSibling` will be `null`, and will give us an error when we try to access its `textContent`)

wbamberg committed:
Update files/en-us/web/api/mutationrecord/previoussibling/index.md

wbamberg commented on files/en-us/web/api/mutationrecord/oldvalue/index.md:
```diff
@@ -0,0 +1,90 @@
+function logOldValue(mutationRecordArray) {
+  for (const record of mutationRecordArray) {
+    log.textContent = `Old value: ${record.oldValue}`;
+    log.style.cssText = record.oldValue;
```
Oh, neat solution :).

wbamberg commented on files/en-us/web/api/mutationrecord/type/index.md:
```diff
@@ -0,0 +1,109 @@
+
+reset.addEventListener("click", () => self.location.reload());
+
+function logNextSibling(records) {
+  for (const record of records) {
+    log.textContent = `Mutation type: ${record.type}`;
+  }
+}
+
+const observer = new MutationObserver(logNextSibling);
```
```suggestion
const observer = new MutationObserver(logMutationType);
```

wbamberg replied:
```suggestion
function logMutationType(records) {
  for (const record of records) {
    log.textContent = `Mutation type: ${record.type}`;
  }
}

const observer = new MutationObserver(logMutationType);
```

wbamberg committed:
Update files/en-us/web/api/mutationrecord/type/index.md

wbamberg commented on files/en-us/web/api/mutationrecord/nextsibling/index.md:
```diff
@@ -0,0 +1,90 @@
+    if (record.type === "childList") {
+      for (const newNode of record.addedNodes) {
+        counter.textContent = `Node count: ${target.children.length}`;
+        log.textContent = `Next sibling of added node: ${record.nextSibling.textContent}`
```
```suggestion
        log.textContent = `Next sibling of added node: ${record.nextSibling?.textContent}`
```

(in case it is `null`)

wbamberg committed:
Update files/en-us/web/api/mutationrecord/nextsibling/index.md

wbamberg commented on files/en-us/web/api/mutationrecord/attributename/index.md:
````diff
@@ -0,0 +1,98 @@
+#### JavaScript
+
+```js
+const hiMom = document.querySelector("#hiMom");
+const redButton = document.querySelector("#redButton");
+const blueButton = document.querySelector("#blueButton ");
+const whiteButton = document.querySelector("#whiteButton");
+const blackButton = document.querySelector("#blackButton");
+const log = document.querySelector("#log");
+redButton.addEventListener("click", () => {
+  hiMom.classList.add("red");
+  hiMom.classList.remove("blue");
+});
+blueButton.addEventListener("click", () => {
+  hiMom.classList.add("blue");
+  hiMom.classList.remove("red");
+});
+whiteButton.addEventListener("click", () => {
+  hiMom.style.color = "white";
+});
+blackButton.addEventListener("click", () => {
+  hiMom.style.color = "black";
+});
+function logLastAttr(mutationRecordArray) {
+  for (const record of mutationRecordArray) {
+    if (record.type === "attributes") {
+      log.textContent = `Updated attribute name: ${record.attributeName}`;
+    }
+  }
+}
+const observer = new MutationObserver(logLastAttr);
+observer.observe(hiMom, { attributes: true });
````
```suggestion
const hiMom = document.querySelector("#hiMom");
const redButton = document.querySelector("#redButton");
const blueButton = document.querySelector("#blueButton ");
const whiteButton = document.querySelector("#whiteButton");
const blackButton = document.querySelector("#blackButton");
const log = document.querySelector("#log");

redButton.addEventListener("click", () => {
  hiMom.classList.add("red");
  hiMom.classList.remove("blue");
});

blueButton.addEventListener("click", () => {
  hiMom.classList.add("blue");
  hiMom.classList.remove("red");
});

whiteButton.addEventListener("click", () => {
    hiMom.style.color = "white";
});

blackButton.addEventListener("click", () => {
  hiMom.style.color = "black";
});

function logLastAttr(mutationRecordArray) {
  for (const record of mutationRecordArray) {
    if (record.type === "attributes") {
      log.textContent = `Updated attribute name: ${record.attributeName}`;
    }
  }
}

const observer = new MutationObserver(logLastAttr);
observer.observe(hiMom, { attributes: true });
```

(sorry I messed up the formatting here)

wbamberg committed:
Update files/en-us/web/api/mutationrecord/attributename/index.md

wbamberg commented on files/en-us/web/api/mutationrecord/target/index.md:
```diff
@@ -0,0 +1,121 @@
+
+#### Result
+
+{{EmbedLiveSample("Red Div, Blue Div", "", 200)}}
```
I like this much better than my suggestion!

wbamberg reviewed (approved):
Thank you @weebney for all your work on this and patience. I think this is a really nice set of pages!

weebney commented:
Thanks so much for the help and patience @wbamberg, I know it was a complete mess from start to finish! Hope to work with you again in the future!

wbamberg commented:
> Thanks so much for the help and patience @wbamberg, I know it was a complete mess from start to finish! Hope to work with you again in the future!

Not at all! It's not a simple thing to make a substantial update to MDN and I think it is worth taking the time to make it as useful as possible.

schalkneethling commented:
> Thanks so much for the help and patience @wbamberg, I know it was a complete mess from start to finish! Hope to work with you again in the future!

Thank you so much for your valuable contribution! Welcome to the community, we look forward to seeing more of your contributions. 🎉
